client/extensions/woocommerce/state/sites/settings/products/test/reducer.js

```diff
@@ -12,11 +13,118 @@ import {
 	WOOCOMMERCE_SETTINGS_BATCH_REQUEST_SUCCESS,
 	WOOCOMMERCE_SETTINGS_PRODUCTS_REQUEST,
 	WOOCOMMERCE_SETTINGS_PRODUCTS_REQUEST_SUCCESS,
+	WOOCOMMERCE_SETTINGS_PRODUCTS_UPDATE_REQUEST_FAILURE,
+	WOOCOMMERCE_SETTINGS_PRODUCTS_UPDATE_REQUEST_SUCCESS,
 } from 'woocommerce/state/action-types';
-import { LOADING } from 'woocommerce/state/constants';
+import { ERROR, LOADING } from 'woocommerce/state/constants';
 import reducer from 'woocommerce/state/sites/reducer';
 
 describe( 'reducer', () => {
+	test( 'should mark the settings products tree as "error" when update request fails', () => {
+		const siteId = 123;
+		const action = {
+			type: WOOCOMMERCE_SETTINGS_PRODUCTS_UPDATE_REQUEST_FAILURE,
+			siteId,
+		};
+
+		const newSiteData = reducer( {}, action );
+		expect( newSiteData[ siteId ].settings.products ).to.eql( ERROR );
+	} );
+
+	test( 'should not change if data.update is missing', () => {
+		const siteId = 123;
+		const settings = [
+			{
+				id: 'some-setting',
+				value: 'yes',
+			},
+			{
+				id: 'another-setting',
+				value: 'no',
+			},
+			{
+				id: 'chicken-and-ribs',
```

Comment:
Mmmmmm :)

Reply:
ha - kind of a meme from i/o, and i carry it on to all test suites wide and far /cc @aduth